packages/components/CONTRIBUTING.md

```diff
@@ -180,6 +183,65 @@ All new component should be styled using [Emotion](https://emotion.sh/docs/intro
 
 Note: Instead of using Emotion's standard `cx` function, the custom [`useCx` hook](/packages/components/src/utils/hooks/use-cx.ts) should be used instead.
 
+
+### Deprecating styles
+
+Changing the styles of a non-experimental component must be done with care. To prevent serious breakage in third-party usage, in some cases we may want a grace period before fully removing the old styles. This can be done by temporarily placing the new styles behind a feature flag prop prefixed by `__next`, accompanied by a `deprecate()` warning in the console. The feature flag should be opt-in (false by default), and have a reasonably descriptive name (**not** `__nextHasNewStyles`). A descriptive name allows for multiple deprecations to proceed in parallel, separated by concerns or by deprecation version.
+
+```jsx
+// component.tsx
+import deprecated from '@wordpress/deprecated';
+import { Wrapper } from './styles.ts';
+
+function MyComponent({ __nextHasNoOuterMargins = false }) {
+	if ( ! __nextHasNoOuterMargins ) {
+		deprecated( 'Outer margin styles for wp.components.MyComponent', {
+			since: '6.0',
+			version: '6.2', // Set a reasonable grace period depending on impact
+			hint:
+				'Set the `__nextHasNoOuterMargins` prop to true to start opting into the new styles, which will become the default in a future version.',
+		} );
+	}
+	return <Wrapper __nextHasNoOuterMargins={__nextHasNoOuterMargins} />
+}
+```
+
+Styles should be structured so the deprecated styles are cleanly encapsulated, and can be easily removed when the deprecation version arrives.
+
+```js
+// styles.ts
+const deprecatedMargins = ({ __nextHasNoOuterMargins }) => {
+  if ( ! __nextHasNoOuterMargins ) {
+    return css`
+      margin: 8px;
+    `;
+  }
+};
+
+export const Wrapper = styled.div`
+  margin: 0;
+
+  ${deprecatedMargins}
+`;
+```
+
+Once deprecated, code examples in docs/stories should include the opt-in prop set to `true` so that new consumers are encouraged to adopt it from the start.
```

Comment:
To explain the procedure more explicitly, we could mention that after the grace period is over, the `__next` property (with its deprecation notice) should be removed from the code completely and that the new styles/behaviour should be always applied.

Reply:
Good idea 👍